sycl/doc/extensions/proposed/sycl_ext_oneapi_clock.asciidoc

```diff
@@ -0,0 +1,176 @@
+not rely on APIs defined in this specification.*
+
+== Backend support status
+
+The APIs in this extension may be used only on a device that has
+`aspect::ext_oneapi_clock`.  The application must check that the device has
+this aspect before submitting a kernel using any of the APIs in this
+extension.  If the application fails to do this, the implementation throws
+a synchronous exception with the `errc::kernel_not_supported` error code
+when the kernel is submitted to the queue.
+
+== Overview
+
+This extension introduces a new free function `clock(scope)`. This function
+allows the user to sample the value from one of three clocks provided by the
+compute units, depending on the value of the scope argument. `scope` is an
+enumeration constant of the new `clock_scope` enum. It should be passed to the
+function to define the clock source; e.g., `clock(clock_scope::sub_group)`
+samples the value from a clock shared by all work-items executing in the same
+sub-group.
+
+This extension also adds a new aspect, `ext_oneapi_clock`, indicating whether
+the feature is supported by the device.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_CLOCK` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the feature-test
+ macro always has this value.
+|===
+
+=== New device aspect
+
+This extension adds new device aspect:
+
+```c++
+namespace sycl {
+
+enum class aspect : /*unspecified*/ {
+  ext_oneapi_clock
+};
+
+} // namespace sycl
+```
+
+[width="100%",%header,cols="50%,50%"]
+|===
+|Aspect
+|Description
+
+|`ext_oneapi_clock`
+|Indicates that the device supports the `sycl::ext::oneapi::experimental::clock()` call.
+|===
+
+=== New enum
+
+```c++
+namespace sycl::ext::oneap::experimental {
+
+enum class clock_scope : /* unspecified */ {
+  sub_group,
+  work_group,
+  device
+}
+
+}; // namespace sycl::ext::oneap::experimental
+```
+An enumerator from `clock_scope` passed to the `clock()` function defines the
+clock source:
+
+[width="100%",%header,cols="50%,50%"]
+|===
+|Enumerator
+|Description
+
+|`sub_group`
+|`clock()` gets values shared by all work-items executing in the same sub-group.
+
+|`work_group`
+|`clock()` gets values shared by all work-items executing in the same work_group.
+
+|`device`
+|`clock()` gets values shared by all work-items executing on the device.
+|===
+
+=== New free function
+
+```c++
+namespace sycl::ext::oneap::experimental {
+
+uint64_t clock(clock_scope scope = clock_scope::sub_group);
```

Comment:
is it possible to use the clock() to implement the usleep(usecond) function in kernel?  or just provide a device function usleep()?  Thanks.

Reply:
There is no way to convert these clocks to seconds, so unfortunately the answer is no.